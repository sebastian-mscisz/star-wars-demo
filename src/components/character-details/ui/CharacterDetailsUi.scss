@import 'src/variables';

.sw-character-detail {
    font-size: 0.9rem;
    color: $base-color;
    text-align: center;
    width: 80%;
    max-width: 960px;
    margin-top: 15px;
    background-color: $base-background-color;
    border-radius: 10px;
    border: 2px solid $base-color;
    display: flex;
    flex-wrap: wrap;

    &__name {
        font-size: 1rem;
        width: 40%;

        &__image-wrapper {
            display: flex;
            width: 100%;
            aspect-ratio: 8/11;
            border-radius: 10px;

            & img {
                width: 100%;
                height: 100%;
                border-radius: 10px;
                outline: 2px solid $base-color;
                aspect-ratio: 8/11;
            }
        }
    }

    &__details {
        text-align: left;
        margin: 10px 25px;
        flex: 1;

        &__link {
            text-decoration: underline;
            font-weight: 700;
            color: $base-color;

            &:hover {
                color: $base-color-light;
            }
        }

        & hr {
            width: 100%;
        }
    }
}
