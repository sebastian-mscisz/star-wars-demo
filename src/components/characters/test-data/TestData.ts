import { Characters, Character, Species } from 'src/common';

export const TestCharacter: Character = {
    name: 'Character1',
    url: 'https://swapi.info/api/people/1',
    homeworld: 'https://swapi.info/api/planets/1',
    species: ['species/1'],
    vehicles: ['https://swapi.info/api/vehicles/1'],
};

export const TestCharacters: Characters = [
    TestCharacter,
    {
        name: 'Character2',
        url: 'https://swapi.info/api/people/2',
        vehicles: ['https://swapi.info/api/vehicles/1', 'https://swapi.info/api/vehicles/2'],
        homeworld: 'https://swapi.info/api/planets/2',
        species: ['species/2'],
    },
    {
        name: 'Character3',
        url: 'https://swapi.info/api/people/3',
        vehicles: [],
        homeworld: 'https://swapi.info/api/planets/3',
        species: ['species/3'],
    },
    {
        name: 'Character4',
        url: 'https://swapi.info/api/people/4',
        vehicles: ['https://swapi.info/api/vehicles/3'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/4'],
    },
    {
        name: 'Character5',
        url: 'https://swapi.info/api/people/5',
        vehicles: ['https://swapi.info/api/vehicles/4', 'https://swapi.info/api/vehicles/5'],
        homeworld: 'https://swapi.info/api/planets/4',
        species: ['species/5'],
    },
];

export const TestSpecies: Species = [
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species2',
        url: 'species/2',
    },
    {
        name: 'Species3',
        url: 'species/3',
    },
    {
        name: 'Species4',
        url: 'species/4',
    },
    {
        name: 'Species5',
        url: 'species/5',
    },
];
