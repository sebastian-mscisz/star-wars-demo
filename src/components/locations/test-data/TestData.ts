import { Locations, Location } from 'src/common';

export const TestLocation: Location = {
    name: 'Location1',
    url: 'https://swapi.info/api/planets/1',
    residents: ['https://swapi.info/api/people/1'],
    population: '1000',
};

export const TestLocations: Locations = [
    TestLocation,
    {
        name: 'Location2',
        url: 'https://swapi.info/api/planets/2',
        residents: ['https://swapi.info/api/people/2'],
        population: '5343',
    },
    {
        name: 'Location3',
        url: 'https://swapi.info/api/planets/3',
        residents: ['https://swapi.info/api/people/3'],
        population: '13333',
    },
    {
        name: 'Location4',
        url: 'https://swapi.info/api/planets/4',
        residents: ['https://swapi.info/api/people/4', 'https://swapi.info/api/people/5'],
        population: '9999',
    },
    {
        name: 'Location5',
        url: 'https://swapi.info/api/planets/5',
        residents: [],
        population: '999',
    },
];
