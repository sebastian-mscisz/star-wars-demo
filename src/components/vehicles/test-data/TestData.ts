import { Vehicles, Vehicle } from 'src/common';

export const TestVehicle: Vehicle = {
    name: 'Vehicle1',
    url: 'https://swapi.info/api/vehicles/1',
    vehicle_class: 'Air speeder',
    pilots: ['character/1', 'character/2'],
};

export const TestVehicles: Vehicles = [
    TestVehicle,
    {
        name: 'Vehicle2',
        url: 'https://swapi.info/api/vehicles/2',
        pilots: ['character/2'],
        vehicle_class: 'Land speeder',
    },
    {
        name: 'Vehicle3',
        url: 'https://swapi.info/api/vehicles/3',
        pilots: ['character/4'],
        vehicle_class: 'Water speeder',
    },
    {
        name: 'Vehicle4',
        url: 'https://swapi.info/api/vehicles/4',
        pilots: ['character/5'],
        vehicle_class: 'Fire speeder',
    },
    {
        name: 'Vehicle5',
        url: 'https://swapi.info/api/vehicles/5',
        pilots: ['character/5'],
        vehicle_class: 'Air speeder',
    },
];
