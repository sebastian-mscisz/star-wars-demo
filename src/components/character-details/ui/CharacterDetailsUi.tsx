import React, { useState } from 'react';
import {
    CharacterWithNeighbours,
    Locations,
    Species,
    Vehicles,
    characterDetailsPagePath,
    getIdFromUrl,
    locationDetailsPagePath,
    vehicleDetailsPagePath,
} from 'src/common';
import './CharacterDetailsUi.scss';
import { NavLink } from 'react-router-dom';
import { PaginationButtons, Spinner } from 'src/components';

type CharacterDetailsUiProps = {
    character?: CharacterWithNeighbours;
    characterId?: string;
    locations?: Locations;
    species?: Species;
    vehicles?: Vehicles;
};

export const CharacterDetailsUi: React.FC<CharacterDetailsUiProps> = ({
    character,
    characterId,
    locations,
    species,
    vehicles,
}) => {
    const [imgLoading, setImgLoading] = useState(true);
    if (!character) return null;

    const originWorld = locations?.find((location) => location.url === character.homeworld) || {
        name: 'Unknown origin planet',
    };

    const race = species?.find((el) => el.url === character.species![0])?.name || 'Unknown race';

    const usedVehicles = vehicles
        ?.filter((vehicle) => character.vehicles?.includes(vehicle.url!))
        .map((el, i) => [
            i > 0 && ', ',
            <NavLink
                key={`${el.name} ${i}`}
                className="sw-character-detail__details__link"
                to={vehicleDetailsPagePath(getIdFromUrl(el.url))}
            >
                {el.name}
            </NavLink>,
        ]);

    return (
        <>
            <div className="sw-character-detail">
                <div className="sw-character-detail__name">
                    <div className="sw-character-detail__name__image-wrapper">
                        <div
                            className={`${
                                imgLoading
                                    ? 'sw-character-detail__name__image-wrapper__display-spinner'
                                    : 'sw-character-detail__name__image-wrapper__hide-spinner'
                            }`}
                        >
                            <Spinner />
                        </div>
                        <img
                            className={`${imgLoading ? 'sw-character-detail__name__image-wrapper__hide_img' : ''}`}
                            src={`./characters/${characterId}.jpg`}
                            alt="Star wars character"
                            onLoad={() => setImgLoading(false)}
                        />
                    </div>
                    <h3>{character.name}</h3>
                </div>
                <div className="sw-character-detail__details">
                    <p>
                        <strong>Origin world:</strong>{' '}
                        {originWorld.url ? (
                            <NavLink
                                className="sw-character-detail__details__link"
                                to={locationDetailsPagePath(getIdFromUrl(originWorld.url))}
                                children={originWorld.name}
                            />
                        ) : (
                            originWorld.name
                        )}
                    </p>
                    <hr />
                    <p>
                        <strong>Race:</strong> {race}
                    </p>
                    <hr />
                    <p>
                        <strong>Used vehicles:</strong> {usedVehicles?.length !== 0 ? usedVehicles : 'No used vehicles'}
                    </p>
                </div>
            </div>
            <PaginationButtons
                leftPath={characterDetailsPagePath(getIdFromUrl(character.leftNeighbour.url))}
                rightPath={characterDetailsPagePath(getIdFromUrl(character.rightNeighbour.url))}
                center
            />
        </>
    );
};
