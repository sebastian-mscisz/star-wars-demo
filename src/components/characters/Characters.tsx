import { withHandlers } from 'src/hoc';
import { CharactersUi } from './ui/CharactersUi';
import { useCharacters } from 'src/hooks';

const CharactersWithHandlers = withHandlers(CharactersUi);

export const Characters: React.FC = () => {
    const { data: characters, isLoading, loadingError } = useCharacters();

    return <CharactersWithHandlers isLoading={isLoading} characters={characters} hasError={!!loadingError} />;
};
