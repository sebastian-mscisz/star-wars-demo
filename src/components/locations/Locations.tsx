import { withHandlers } from 'src/hoc';
import { LocationsUi } from './ui/LocationsUi';
import { useLocations } from 'src/hooks';

const LocationsWithHandlers = withHandlers(LocationsUi);

export const Locations: React.FC = () => {
    const { data: locations, isLoading, loadingError } = useLocations();

    return <LocationsWithHandlers isLoading={isLoading} locations={locations} hasError={!!loadingError} />;
};
