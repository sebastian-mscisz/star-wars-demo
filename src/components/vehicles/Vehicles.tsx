import { withHandlers } from 'src/hoc';
import { VehiclesUi } from './ui/VehiclesUi';
import { useVehicles } from 'src/hooks';

const VehiclesWithHandlers = withHandlers(VehiclesUi);

export const Vehicles: React.FC = () => {
    const { data: vehicles, isLoading, loadingError } = useVehicles();

    return <VehiclesWithHandlers isLoading={isLoading} vehicles={vehicles} hasError={!!loadingError} />;
};
