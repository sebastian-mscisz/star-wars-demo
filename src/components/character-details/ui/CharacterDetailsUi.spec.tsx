import { BrowserRouter } from 'react-router-dom';
import { CharacterDetailsUi } from './CharacterDetailsUi';
import { render } from '@testing-library/react';
import { TestCharacters, TestSpecies } from '../../characters/test-data';
import { TestLocations } from '../../locations/test-data';
import { TestVehicles } from '../../vehicles/test-data';

describe('CharacterDetailsUi', () => {
    it('should render', () => {
        const { asFragment } = render(
            <CharacterDetailsUi
                character={{
                    ...TestCharacters[1],
                    leftNeighbour: TestCharacters[0],
                    rightNeighbour: TestCharacters[2],
                }}
                locations={TestLocations}
                species={TestSpecies}
                vehicles={TestVehicles}
                characterId={'3'}
            />,
            { wrapper: BrowserRouter },
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
