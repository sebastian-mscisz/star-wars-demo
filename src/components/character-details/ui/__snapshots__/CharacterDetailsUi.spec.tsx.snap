// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CharacterDetailsUi should render 1`] = `
<DocumentFragment>
  <div
    class="sw-character-detail"
  >
    <div
      class="sw-character-detail__name"
    >
      <div
        class="sw-character-detail__name__image-wrapper"
      >
        <img
          alt="Star wars character"
          src="./characters/3.jpg"
        />
      </div>
      <h3>
        Character2
      </h3>
    </div>
    <div
      class="sw-character-detail__details"
    >
      <p>
        <strong>
          Origin world:
        </strong>
         
        <a
          class="sw-character-detail__details__link"
          href="/location/2"
        >
          Location2
        </a>
      </p>
      <hr />
      <p>
        <strong>
          Race:
        </strong>
         Species2
      </p>
      <hr />
      <p>
        <strong>
          Used vehicles:
        </strong>
         
        <a
          class="sw-character-detail__details__link"
          href="/vehicle/1"
        >
          Vehicle1
        </a>
        , 
        <a
          class="sw-character-detail__details__link"
          href="/vehicle/2"
        >
          Vehicle2
        </a>
      </p>
    </div>
  </div>
  <div
    class="sw-pagination-section sw-pagination-section--center"
    data-testid="pagination-section"
  >
    <div
      class="sw-pagination-button "
      data-testid="left-pagination-button"
    >
      ←
    </div>
    <div
      class="sw-pagination-button "
      data-testid="right-pagination-button"
    >
      →
    </div>
  </div>
</DocumentFragment>
`;
